### ISSUE TITLE
RLLib Apex-DQN regression on convergence

### ISSUE THREAD

**Issue by RuofanKong** [ray-project/ray#18034]
### Issue Description
I am upgrading Ray from 0.8.6 to 1.3.0. I tried training CartPole-v0 (OpenAI Gym) with Apex-DQN and the default hyper-parameters, and I noticed that Apex-DQN 1.3.0 is way slower than 0.8.6 on the convergence w.r.t training iterations. Specifically, the mean episode reward mostly reaches 100+ after 4 - 5 training iterations for 0.8.6, while it's still <100 after 15 training iterations for 1.3.0 (each training iteration has 25K samples by default). I also compared the default hyper-parameters between these two versions, the only difference is as below that I don't believe would cause the convergence performance w.r.t training iterations.

```
# for 1.3.0
'intra_op_parallelism_threads': 8, 'inter_op_parallelism_threads': 8
'placement_strategy': 'PACK'
'replay_mode': 'independent'
count_steps_by': 'env_steps'
'replay_sequence_length': 1
```

### System Info
- Ray version: 1.3.0 (compared against v0.8.6)
- Python version: 3.8.5
- Tensorflow version: 2.3.0
- OS: MacOS Catalina 10.15.6

### Repro Step
Run the following code for both Ray 1.3.0 and 0.8.6, and you can consistently get the convergence delta w.r.t training iterations as I described at beginning.

```
import ray
import ray.rllib.agents.dqn.apex as apex
from ray.tune.logger import pretty_print


def train_cartpole() -> None:
    ray.init()
    config = apex.APEX_DEFAULT_CONFIG.copy()
    config["num_gpus"] = 0
    config["num_workers"] = 3
    trainer = apex.ApexTrainer(config=config, env="CartPole-v0")

    for _ in range(1000):
        # Perform one iteration of training the policy with Apex-DQN
        result = trainer.train()
        print(pretty_print(result))


train_cartpole()
```

### Note
I did some investigations on the issue by separating concerns component-by-component. I narrowed it down and finally confirmed the issue is on the sampler (rllib/evaluation/sampler.py) working with the new trajectory view API. What I did is, I ran the above code but replaced the sampler code with the one in Ray 0.8.6 (and with some minor tuning on the sampler interface to be compatible), while kept all other components running 1.3.0. It could achieve the very close convergence performance results compared against 0.8.6.

**Comment by sven1977** [comment]
Hey @RuofanKong , I can confirm that the latest master is slower than 0.8.6 on your example script. Roughly 10% for Atari and ~20% for CartPole :/ will try to find out why.

**Comment by sven1977** [comment]
Hey @RuofanKong , I did some digging and it's not necessarily the trajectory view API per se. It does still run "fast" (similar to 0.8.6) in versions 1.0.0 and 1.1.0, which both already have the trajectory view API enabled for APEX. The slow-down happens after 1.2.0. I'm currently bisecting to find the actual root cause. ...

**Comment by RuofanKong** [comment]
Hi @sven1977, Thanks for looking into this! Actually perhaps you tried the latest master, what i see is, 1.3.0 consistently got 2x+ slower down on CartPole, compared against 0.8.6. Also as I shared, the issue is not necessarily trajectory API itself, could be sampler, SimpleListCollector, etc. I confirmed it because as I mentioned, If I replaced the 1.3.0 sampler with 0.8.6 one but keep all other components using 1.3.0., the convergence is fine. So I think something on sampler and components interacted with it or so. But yes, need to further root cause it.

**Comment by sven1977** [comment]
Hey @RuofanKong , thanks for the analysis. I got the PR that's responsible for this and yes, it's where we switch on the traj. view API for DQN (and hence for APEX). It's strange, DQN and SimpleQ don't seem affected by this, but I'll take a closer look now. Could have to do with APEX' distributed/sharded buffers. ...

**Comment by sven1977** [comment]
Here is a PR that hopefully fixes the issue. I'll confirm now running the above experiment again.

https://github.com/ray-project/ray/pull/18358

**Comment by sven1977** [comment]
@RuofanKong ^

**Comment by RuofanKong** [comment]
@sven1977 I ran multiple times again with the fix also applied to 1.3.0, now looks it gets the comparable results with 0.8.6 as below as an example. Thanks for the fix!! Close the ticket now.

![Screen Shot 2021-09-06 at 6 09 21 PM](https://user-images.githubusercontent.com/7396554/132269775-5e28baab-59e7-49e8-9fe5-9dcbbe053fbb.png)
